opds2_import.py

```diff
@@ -0,0 +1,830 @@
+            )
+
+        return medium
+
+    def _extract_publication_metadata(self, feed, publication, data_source_name):
+        """Extract a Metadata object from webpub-manifest-parser's publication.
+
+        :param publication: Feed object
+        :type publication: opds2_ast.OPDS2Feed
+
+        :param publication: Publication object
+        :type publication: opds2_ast.OPDS2Publication
+
+        :param data_source_name: Data source's name
+        :type data_source_name: str
+
+        :return: Publication's metadata
+        :rtype: Metadata
+        """
+        self._logger.debug(
+            u"Started extracting metadata from publication {0}".format(
+                encode(publication)
+            )
+        )
+
+        title = publication.metadata.title
+
+        if title == OPDSFeed.NO_TITLE:
+            title = None
+
+        subtitle = publication.metadata.subtitle
+
+        languages = first_or_default(publication.metadata.languages)
+        derived_medium = self._extract_medium_from_links(publication.links)
+        medium = self._extract_medium(publication, derived_medium)
+
+        publisher = first_or_default(publication.metadata.publishers)
+        if publisher:
+            publisher = publisher.name
+
+        published = publication.metadata.published
+        subjects = self._extract_subjects(publication.metadata.subjects)
+        contributors = (
+            self._extract_contributors(publication.metadata.authors)
+            + self._extract_contributors(publication.metadata.translators)
+            + self._extract_contributors(publication.metadata.editors)
+            + self._extract_contributors(publication.metadata.artists)
+            + self._extract_contributors(publication.metadata.illustrators)
+            + self._extract_contributors(publication.metadata.letterers)
+            + self._extract_contributors(publication.metadata.pencilers)
+            + self._extract_contributors(publication.metadata.colorists)
+            + self._extract_contributors(publication.metadata.inkers)
+            + self._extract_contributors(publication.metadata.narrators)
+            + self._extract_contributors(publication.metadata.contributors)
+            + self._extract_contributors(publication.metadata.imprints)
+        )
+
+        feed_self_url = first_or_default(
+            feed.links.get_by_rel(OPDS2LinkRelationsRegistry.SELF.key)
+        ).href
+        links = self._extract_links(publication, feed_self_url)
+
+        last_opds_update = publication.metadata.modified
+
+        identifier, _ = Identifier.parse_urn(self._db, publication.metadata.identifier)
+        identifier_data = IdentifierData(
+            type=identifier.type, identifier=identifier.identifier
+        )
+
+        # FIXME: There are no measurements in OPDS 2.0
+        measurements = []
+
+        # FIXME: There is no series information in OPDS 2.0
+        series = None
+        series_position = None
```

Comment:
This seems like something we need to work out with Hadrien, probably using https://schema.org/CreativeWorkSeries.

Reply:
Actually I found an example in Feedbooks feed:
```
"belongsTo": {
          "collection": "The Age of Madness",
          "series": [{
            "name": "The Age of Madness",
            "position": 2,
            "links": [{
              "type": "application/opds+json",
              "href": "https://catalog.feedbooks.com/series/32576.json?lang=en"
            }]
          }]
        },
```

Reply:
Yes this is covered in RWPM and OPDS 2.0: https://readium.org/webpub-manifest/contexts/default/#collections--series